### ISSUE TITLE
[ENH] Improve the test framework of v1 and make it more robust

### ISSUE THREAD

**Issue by phoeenniixx** [sktime/pytorch-forecasting#1906]
**Is your feature request related to a problem? Please describe.**
The current test framework for model testing in v1 is inspired from the previous "hard-coded" tests for the models and needs some improvements to test which models are compatible with what kind of `loss` functions. 
Right now:
- The output format of the `forward` pass depends on the type of loss being used.
- There is no concrete contract defining the expected output format for each model-loss combination.
- This makes it difficult to know the model behavior and leads to fragile tests.

Moreover, right now, not all models in the package are tested through this test framework, so we need to add all the remaining models to the test framework as well.

**Describe the solution you'd like**
We define  a tag `info:compatible_loss` (or any other better name?).
This tag is a list of strings, each representing a class of loss functions the model supports:
*  `"distr"` - Distribution Losses like `NormalDistributionLoss`
* `"quantile"` - `QuantileLoss`
* `"point"` - Point Prediction Losses like `MAE`

The tag for each model will list one or more of these strings, indicating compatible loss classes. For example:

```python
"info:compatible_loss" = ["point"]
```
means the model is compatible with point prediction losses, but not distribution losses and quantile losses.

Test Behavior:

- For each compatible loss class, the model will be tested with all applicable losses in that class to catch potential "edge cases".
- Each test will also verify that the output format (2D/3D tensor, correct shape) matches expectations.

Special case — `MultiLoss`:
Based on current manual testing, we assume that all models are compatible with `MultiLoss` (which combines multiple loss classes).
- All models will therefore be tested with `MultiLoss`.
- If any model fails when tested with `MultiLoss`, we will introduce an additional string `"multi"` to the tag list, indicating that the model explicitly supports `MultiLoss`.

Optionally, We should also think of a way to directly save the output format for each model-loss (or model-loss class?) pair, to see which models output formats are supported for each pair. This functionality later can be disabled. This could provide some insights about the output formats so that we can decide on eventual unified output format contract for all models.
Any thoughts @fkiraly , @agobbifbk, @PranavBhatP, @xandie985?

**Additional context**
This effort is mainly to understand more about `Metrics` and models and their interactions for the[ design document](https://hackmd.io/DL0_NjbERIy65laob22GlA) on concrete contracts model outputs and metrics.

**Comment by PranavBhatP** [comment]
Can you change the name of the issue to specify testing of models? Just pointing this out since we have two issues working on closely related topics and it might be easier to differentiate?

**Comment by fkiraly** [comment]
> we assume that all models are compatible with `MultiLoss`

I think we need to specify instances of `MultiLoss` for the correct statement, since a `MultiLoss` is composite and constructed with other losses.

I think only instances of `MultiLoss` are possible to use where the component losses belong to "allowed" metrics.

**Comment by phoeenniixx** [comment]
Yes, I have updated the issue, Thanks for spotting it!

**Comment by phoeenniixx** [comment]
> Optionally, We should also think of a way to directly save the output format for each model-loss (or model-loss class?) pair, to see which models output formats are supported for each pair. This functionality later can be disabled. This could provide some insights about the output formats so that we can decide on eventual unified output format contract for all models.

Do we need such kind of methods? Any ideas? @fkiraly @PranavBhatP @agobbifbk

**Comment by fkiraly** [comment]
Can you be more specific on the functionality that you have in mind, @phoeenniixx?

Getting the output from *before* the loss is applied?

**Comment by phoeenniixx** [comment]
Well As we know different models behave differently with each loss:
- `TFT` returns a 3D tensor even in case of point prediction losses
- While `TimeXer` returns a 2D tensor for point prediction losses and 3D tensor for other losses.
- `NBEATS` returns 2D always

Because of this variability, a simple test that checks if the output is 2D or 3D is not sufficient. What we actually need is a systematic way to record and track.
We need this data- which model returns what shape of output with each loss, so that we can understand the working and can decide what to do in future. 
So, What I am saying is can we add some methods that not only checks the shape of the output tensor of `forward` but also saves this data in this format:
```python
<MODEL> - <LOSS CLASS> - <OUTPUT SHAPE>
```
so that we have this data and we can take further actions?
Or we can do it in some other way?

This functionality can be disabled once we have the whole data and we have a concrete plan?

**Comment by xandie985** [comment]
> * If any model fails when tested with `MultiLoss`, we will introduce an additional string `"multi"` to the tag list, indicating that the model explicitly supports `MultiLoss`.

@phoeenniixx  if the model "passes" a particular loss, then we add that loss to the list. right?

**Comment by phoeenniixx** [comment]
> > * If any model fails when tested with `MultiLoss`, we will introduce an additional string `"multi"` to the tag list, indicating that the model explicitly supports `MultiLoss`.
> 
> [@phoeenniixx](https://github.com/phoeenniixx) if the model "passes" a particular loss, then we add that loss to the list. right?

Not exactly, if the model passes ALL the losses of a particular loss class, then we add that loss class to the list
there are 3 loss classes:
- `"distr"` - Distribution Losses like `NormalDistributionLoss`
- `"quantile"` - `QuantileLoss`
- `"point"` - Point Prediction Losses like `MAE`

Here for now, we are assuming if a model passes one loss of a loss class, it will pass all the losses of that particular class. To test this exact assumption, we are looping over all the losses of a specific loss class in #1908 and checking if this assumption is actually true or not.
